File: arraysandbinarysearch/bs.java
import java.util.Arrays;

public class bs{

    public static void main(String[] args){
        System.out.println("hola mundial");
        int[] arr = new int[new Double(Math.floor(Math.random()*1000)).intValue()];
        int length = arr.length;
        for(int i= 0; i < length; i++){
            arr[i] = new Double(Math.floor(Math.random()*length)).intValue();
        }

        double flooredIndex = Math.floor(Math.random()*length);
        int flooredIndexInt = new Double(flooredIndex).intValue();
        int target = arr[flooredIndexInt];
        System.out.println("searching for " + target);
        Arrays.sort(arr);
        int javaAnswer = Arrays.binarySearch(arr,target);
        System.out.println( javaAnswer + " java answer" );
        System.out.println(arr[javaAnswer]);
        System.out.println(arr[0] + " "  + arr[length-1]);
        
        int left = 0;
        int right = length -1;
        int middle = -1;
        int iterations = 0;
        while(left <= right){
            
            middle = (left+right)/2;
            System.out.println(left + " " + middle + " " + right + " " + target + " " + flooredIndexInt);
            if(arr[middle] == target){
                System.out.println(middle);
                break;
            }
            if(arr[middle] > target){
                right = middle - 1;
            }else if(arr[middle] < target){
                left = middle + 1;
            }
            iterations++;
            
        }

        System.out.println( arr[middle] + " " + target);
        // System.out.println(Math.(length) + " log of length");
        System.out.println(iterations +" iterations");
    }
}

File: arraysandbinarysearch/bs.kt
import java.util.Arrays
import java.util.Random
import java.time.LocalDateTime
import java.time.ZoneOffset
import java.time.Instant

fun main(args: Array<String>) {
    // a new random generator
    val random = Random(Instant.now().toEpochMilli());
    var randLength = random.nextInt(1000);
    
    // random length
    var x =  IntArray(randLength,{i -> random.nextInt(1000)})
    x.sort()

    // random target
    val target = x[random.nextInt(randLength)];
    var left = 0
    var right = randLength-1
    var middle: Int =0

    // print the target
    println("$target is the target")

    // loop until we converge
    while (left <= right){

        // define the middle
        middle = ( (left + right)/2 )
        if(x[middle] == target){
            break;
        }

        if(x[middle] > target){
            right = middle - 1;
        }else{
            left = middle + 1
        }
    }

    // print what we have
    val mid = x.get(middle)
    println("$target and $mid")
}